tools/cli/NearObjectCli.cxx

```diff
@@ -227,12 +227,25 @@ NearObjectCli::AddSubcommandUwbRangeStart(CLI::App* parent)
 
     // TODO is there a way to put all the enums into a list of [optionName, optionDestination, optionMap] so we don't have to create the initializer list each time
 
-    // enumerations
+    // List mandatory params first
+    // TODO: --help should output that these params are mandatory
     detail::AddEnumOption(rangeStartApp, uwbConfig.deviceRole);
+    detail::AddEnumOption(rangeStartApp, uwbConfig.multiNodeMode);
+    rangeStartApp->add_option("--NumberOfControlees", uwbConfig.numberOfControlees, "1 <= N <= 8")->capture_default_str(); // TODO: Input validation
```

Comment:
I'm wondering if we care to specify this at all. If we allow the user/caller to specify N controlees, then we'll know exactly how many controlees there are. My understanding is that if you specify `N` here, then `N` controlee addresses need to be specified as well, which would make this redundant. Thoughts?

Reply:
I think we still need this, only for the case when the user is configuring a controlee device in a multicast session. In this scenario, DeviceType will be Controlee, DeviceMacAddress (from UCI perspective) will be this particular device's mac address, and DstMacAddress (from UCI perspective) will be the Controller's mac address. So the user isn't going to input the mac addresses of ALL controlees, but the number of controlees is still needed.

However, now that I type this out, I'm realizing that I was originally intending to allow the user of the tool to specify the Controller and N Controlees, which conflicts with what I just said. Do you think there is value in the user of this tool to have to supply all of the information (meaning specify the Controller and all N Controlees), or should the user only care about their own device (meaning specify the Controller and itself, plus number of controlees)?

Reply:
> Do you think there is value in the user of this tool to have to supply all of the information (meaning specify the Controller and all N Controlees), or should the user only care about their own device (meaning specify the Controller and itself, plus number of controlees)?

I think your interpretation is correct as the `NUMBER_OF_CONTROLEES` shows up in CSML `UWB_CONFIGURATION` where `DST_MAC_ADDRESS` isn't described at all (only the controlee short mac and controller mac, both singular, show up):

![image](https://user-images.githubusercontent.com/2082148/226184324-89d34789-23b6-48f8-b569-b86b7a82c3b4.png)
![image](https://user-images.githubusercontent.com/2082148/226183118-478055c1-e488-4fe7-a999-485b5e0a2035.png)

meaning the number of controlees can't be tied to `DST_MAC_ADDRESSES` and so it makes sense that there's a separate application configuration parameter for it.

Consequently, I think we have it encoded wrong in the CLI. We should allow specifying the "self" mac address (`DeviceMacAddress`) which is only ever singular, and then the "destination" mac addresses, of which there can be many. For the self mac address, we should also allow an option to generate a random address (`UwbMacAddress::Random()` supports this). I also agree with you that specifying all the controlee addresses should only be needed when the host is a controller:

| Device Type  | Device Mac Address | Destination Mac Addresses |
| --- |  --- | --- |
| Controller | Own Address  | 1<= N <= 8 Controlee Addresses |
| Controlee | Own Address  | Controller Address |

Reply:
I agree about the encoding, so I will change the CLI to have an option for DeviceMacAddress (referring to "self" mac address), as well as DestinationMacAddresses (either controller is host is a controlee, or all controlees if host is controller). Unfortunately, I would have to list all the options all at once due to the nature of command line arguments. Ideally, we would first see what the DeviceType is (controller/controlee), then display the other arguments depending on that value, but I don't think that's possible